src/index.js

```diff
@@ -1 +1,34 @@
-export default () => true
+import defaultsDeep from 'lodash.defaultsdeep';
+import pickBy from 'lodash.pickby';
+import makeOkCheck from './make-ok-check';
+import buildOptions from './build-options';
+import resolveUrl from './resolve-url';
+
+class Stipulate {
+
+  constructor(options) {
+    this.baseOptions = options;
+  }
+
+  beforeRequest(url, options) {
+    return [url, options];
+  }
+
+  afterResponse(response) {
+    return response;
+  }
+
+  send(urlString, options, query) {
+    const config = buildOptions(options, this.baseOptions);
+    const url = resolveUrl(urlString, query);
+
+    const requestArguments = this.beforeRequest(url, config);
+    const checkOk = makeOkCheck(config);
+
+    return fetch(...requestArguments)
+      .then(checkOk)
+      .then(this.afterResponse);
```

Comment:
>The isolation is good because the user's code doesn't interact directly with the internal code and vice versa.

I'm not sure what you mean by "interact directly", because by my definition it does exactly that, just via a cup-game of function calls.

>I added my own before and after logic that gets executed without changing the doSomething function and it doesn't affect the user's code either.

the problem with your examples is they are function calls that, presumably, have some sort of side effect that is indirectly important is some way. (i.e. you're not using the return value of `doSomethingElseBeforeDoSomething()`) That being the case you can't say that it didn't affect the user's code, it just didn't _noticeably_ affect it. who knows what it did via side effects. (because presumably it's not a no-op).

>I added my own before and after logic that gets executed without changing the doSomething function

that's achievable, and more flexible, with the current implementation and using `super`, because whatever changes i want to make can be in the before and after stub functions, and users that extend to override those functions can choose to `super` in my additions or not, at their discretion.

>It's a subtle difference for stronger code.

I don't see the strength you're referring to. I see extraneous type-checking and masking of functions to keep users from taking advantage of class inheritance, which is part of es2015.

>I also think it's a lot cleaner of an implementation.

I think it's the exact opposite. 😦